CIS-4: Credential Registry Standard (#45)
Concordium/concordium-update-proposals: source/CIS/cis-4.rst

It is serialized as :ref:`CIS-4-ContractAddress` (``contract_address``), :ref:`CIS-4-EntrypointName` (``entry_point``), :ref:`CIS-4-Nonce` (``nonce``), and :ref:`CIS-4-Timestamp` (``timestamp``)::

  SigningData ::= (contract_address: ContractAddress) (entry_point: EntrypointName) (nonce: Nonce) (timestamp: Timestamp)

.. _CIS-4-SchemaRef:

``SchemaRef``
^^^^^^^^^^^^^

A URL of the credential schema.

Serialized in the same way as :ref:`CIS-2 MetadataUrl<CIS-2-MetadataUrl>`.

.. _CIS-4-CredentialType:

``CredentialType``
^^^^^^^^^^^^^^^^^^

Is an short ASCII string (up to 256 characters) describing the credential type that is used to identify which schema the credential is based on.
It corresponds to a value of the ``name`` attribute of the credential schema.

It is serialized as: First byte encodes the length (``n``) of the credential type, followed by this many bytes for the credential type string::

  CredentialType ::= (n: Byte) (credential_type: Byteⁿ)

.. _CIS-4-Commitment:

``Commitment``
^^^^^^^^^^^^^^

> A vector Pedersen commitment to the credential attributes.

Author: This was one of the possibilities, but in the science team we picked "vector Pedersen commitment" and it is used in this form in the blue paper. Let's stick with it here as well.

> A vector Pedersen commitment to the credential attributes.

Reviewer: Ok, fair enough. 👍

- :ref:`CIS-4-functions-credentialStatus`
- :ref:`CIS-4-functions-issuer`
- :ref:`CIS-4-functions-issuerMetadata`
- :ref:`CIS-4-functions-registerCredential`
- :ref:`CIS-4-functions-revokeCredentialIssuer`
- :ref:`CIS-4-functions-revokeCredentialHolder`
- :ref:`CIS-4-functions-revokeCredentialOther`
- :ref:`CIS-4-functions-registerRevocationKeys`
- :ref:`CIS-4-functions-removeRevocationKeys`
- :ref:`CIS-4-functions-revocationKeys`

.. _CIS-4-functions-credentialEntry:

``credentialEntry``
^^^^^^^^^^^^^^^^^^^

Query a credential entry from the registry by ID.

Parameter
~~~~~~~~~

The parameter is the credential ID.

See the serialization rules in :ref:`CIS-4-CredentialID`.

Response
~~~~~~~~

The function returns a registry entry corresponding to the credential ID parameter.

It is serialized as :ref:`CIS-4-CredentialInfo` (``credential_info``) followed by a credential schema reference :ref:`CIS-4-SchemaRef` (``schema_ref``), and a credential entry revocation nonce :ref:`CIS-4-Nonce` (``revocation_nonce``)::

  CredentialQueryResponse ::= (credential_info: CredentialInfo) (schema_ref: SchemaRef) (revocation_nonce: Nonce)

Requirements
~~~~~~~~~~~~

> - The query MUST fail if the credential ID is unknown.

Reviewer: ```suggestion
- The query MUST fail if the credential ID is not present in the registry.
```
Reuse the phrasing from further below. `unknown` can be ambiguous.

> - The query MUST fail if the credential ID is unknown.

Author: > Should we also state that it must return the info if the id is present?

I don't think we need to explicitly say it in `Requirements`, the `Response` section states what is returned.

> - The query MUST fail if the credential ID is unknown.

Author: I'll change the phrasing "unknown" -> "not present in the registry" in all the relevant cases.

``credentialStatus``
^^^^^^^^^^^^^^^^^^^^^^^^

Query the status of a credential from the credential registry by ID.

Parameter
~~~~~~~~~

The parameter is the credential ID.

See the serialization rules in :ref:`CIS-4-CredentialID`.

Response
~~~~~~~~

The function returns the status of a credential.

See the serialization rules in :ref:`CIS-4-CredentialStatus`

Requirements
~~~~~~~~~~~~

- The query MUST fail if the credential ID is unknown.
- The credential status MUST be ``Expired`` if the credential is not revoked, the field ``valid_until`` was present in :ref:`CIS-4-CredentialInfo` when registering the credential, and ``valid_until < now``.
- The credential status MUST NOT be ``Expired`` if the field ``valid_until`` was not present in :ref:`CIS-4-CredentialInfo` when registering the credential.
- The credential status MUST be ``Acive`` if the credential is not revoked, and does not qualify as ``Expired`` or ``NotActivated``.

.. _CIS-4-functions-issuer:

``issuer``
^^^^^^^^^^

Query the issuer's account address.

Response
~~~~~~~~

The function output is the issuer's account address.
It is serialized as :ref:`CIS-4-AccountAddress`.

.. _CIS-4-functions-issuerMetadata:

> ``issuerMetadata``

Author: This entrypoint doesn't have a parameter. This is metadata for the whole register, since it's a single issuer per register.
(There is a different kind of metadata: credential metadata. To query that one you need a credential id)

> ``issuerMetadata``

Reviewer: Ah, my mistake.

^^^^^^^^^^^^^^^^^^

Query the current token metadata URLs for a list of token IDs.

Response
~~~~~~~~

The function output is the issuer's metadata URL.

> It is serialized as :ref:`CIS-2-MetadataUrl`.

Author: No, it's just a single `MetadataUrl` value for the issuer.

> It is serialized as :ref:`CIS-2-MetadataUrl`.

Reviewer: Ok

.. _CIS-4-functions-registerCredential:

``registerCredential``
^^^^^^^^^^^^^^^^^^^^^^

Register a credential with the given ID.

Parameter
~~~~~~~~~

The parameter is the credential ID and credential information that is used to create an entry in the registry.

It is serialized as :ref:`CIS-4-CredentialID` (``credential_id``) followed by :ref:`CIS-4-CredentialInfo` (``credential_info``)::

  RegisterCredentialParameter ::= (credential_id: CredentialID) (credential_info: CredentialInfo)

Requirements
~~~~~~~~~~~~

- The credential registration request MUST fail if the credential ID is already present in the registry.

> - The credential status after registration MUST NOT be ``Revoked`` (See the possible values for the status in :ref:`CIS-4-CredentialStatus`).

Author: Fixed.

.. _CIS-4-functions-revokeCredentialIssuer:

``revokeCredentialIssuer``
^^^^^^^^^^^^^^^^^^^^^^^^^^

Revoke a credential by the issuer's request.
The issuer is authorized to revoke a credential if the transaction sender's address is the same as the return value of :ref:`CIS-4-functions-issuer`.

Parameter
~~~~~~~~~

The parameter is the credential ID :ref:`CIS-4-CredentialID` and optional string indicating the revocation reason.

It is serialized as :ref:`CIS-4-CredentialID` followed by 1 byte to indicate whether a reason is included, if its value is 0, then no reason string present, if the value is 1 then the bytes corresponding to the reason string follow::

  OptionReason ::= (0: Byte)
                 | (1: Byte) (n: Byte) (reason_string: Byteⁿ)
  RevokeCredentialIssuerParam ::= (credential_id: CredentialID) (reason: OptionReason)

> .. TODO: what kind of characters are allowed? ASCII, Unicode?

Author: Fixed TODO, use UTF-8.

Requirements
~~~~~~~~~~~~

- If revoked successfully, the credential status MUST change to ``Revoked`` (see :ref:`CIS-4-functions-credentialStatus`).
- The revocation MUST fail if:
    - The the transaction ``sender`` is not the issuer.

>     - The credential ID is not present in the registry.

Author: I've considered this in the beginning. After discussing it with @abizjak, it looks like requiring certain rejection codes restricts implementation. One needs to specify the exact situations when the codes are returned in order for this to be reliably used in the wallet.
After that, I've decided to not specify any rejection codes.

>     - The credential ID is not present in the registry.

Reviewer 2: This is based on the experience of using CIS2 contracts.

The codes are not used because they are not reliable in general. So it's hard to make use of them in a meaningful way, e.g., in a client that interacts with general CIS2 contracts.

>     - The credential ID is not present in the registry.

Reviewer 2: The issue being that the codes are only defined in the sense of "if code is X then this must have happened", but we do not prescribe an `if and only if` since it prescribes implementation too much. For example our contracts often fail with "serialization failure".

>     - The credential ID is not present in the registry.

Reviewer 3: I agree, that we don't want to add requirements for the error codes, but standardizing an error code and the meaning of it, is still useful. If you want your contract to give a good user experience, you would want to provide the right error information back to the user and having some standardized error codes for this helps.

>     - The credential ID is not present in the registry.

Reviewer 2: What I tried to write above is that I don't think it actually helps.

But if you want to do it I'm not against it, but for me there isn't much value in this rpecisely because you cannot rely on it.

>     - The credential ID is not present in the registry.

Reviewer 3: An application should not rely their logic on this, no. But in the case of a UI it is useful, the wallet could show a nice error message and the contract developer would be motivated to take advantage of this.
I guess the schema for error codes, sort of solves this in the general case.

>     - The credential ID is not present in the registry.

Reviewer 3: This is just a consideration

>     - The credential ID is not present in the registry.

Author: It's hard for me to see which codes to standardize and which not. Issuance is usually issuer-specific, not much to standardize there; verifiers and holders (wallets) mostly fetch data, so, the only relevant error is `CredentialNotFound`.
Holders can revoke in the wallet, but it can fail for multiple reasons, and it's hard to fix the granularity in advance.

This essentially leaves us with  `CredentialNotFound`, which I can include in the standard. It would require a slight change in the credential registry example:  the order of enum constructors [here](https://github.com/Concordium/concordium-rust-smart-contracts/blob/main/examples/credential-registry/src/lib.rs#L172) to make the CIS-4 error have the first tag.
And, of course, we should provide a library in the future similar to CIS-2 that will help avoid error code clashes.

@abizjak @limemloh How does that sound?

>     - The credential ID is not present in the registry.

Author: > the order of enum constructors [here](https://github.com/Concordium/concordium-rust-smart-contracts/blob/main/examples/credential-registry/src/lib.rs#L172) to make the CIS-4 error have the first tag.

And I'll probably need to do something similar to the CIS-2 library https://github.com/Concordium/concordium-rust-smart-contracts/blob/main/concordium-cis2/src/lib.rs#L923

>     - The credential ID is not present in the registry.

Author: I experimented with adding a specific rejection code to the implementation and it looks like it brings more issues than solves. One should make sure that for the `CredentialNotFound` case a specific rejection code is produced and all custom errors don't clash with it. Since there is no clear benefit from doing this, I'll drop this idea and settle on not including any rejection code requirements.

    - The credential status is not one of ``Active`` or ``NotActivated`` (see :ref:`CIS-4-functions-credentialStatus`).

.. _CIS-4-functions-revokeCredentialHolder:

``revokeCredentialHolder``
^^^^^^^^^^^^^^^^^^^^^^^^^^

Revoke a credential by the holders's request.

The holder is authorized to revoke a credential by verifying the signature with the holder's public key.

> It replaces the authorization checks conducted on the ``sender/invoker`` variable with signature verification.

Reviewer 3: What is the `sender/invoker` variable? Are you referring to the sender of the contract update? If so, consider making this more explicit, ideally this standard would make sense to people who are not familiar to our smart contract Rust API

Parameter
~~~~~~~~~

It is serialized as :ref:`CIS-4-SignatureEd25519` (``signature``) and message data ``RevocationDataHolder`` consisting of :ref:`CIS-4-CredentialID` (``credential_id``), metadata about the signature :ref:`CIS-4-SigningData` (``signing_data``), and an optional revocation reason (``reason``), serialized similarly to :ref:`CIS-4-functions-revokeCredentialIssuer`::

  RevocationDataHolder ::= (credential_id: CredentialID) (signing_data: SigningData) (reason: OptionReason)
  RevokeCredentialHolderParam ::= (signature: SignatureEd25519) (data : RevocationDataHolder)

Requirements
~~~~~~~~~~~~

- If revoked successfully, the credential status MUST change to ``Revoked`` (see :ref:`CIS-4-functions-credentialStatus`).
- The ``RevokeCredentialHolderParam``'s ``signing_data`` MUST include a nonce to protect against replay attacks.
  The holders's nonce is sequentially increased every time a revocation request is successfully executed.
  The function MUST only accept a ``RevokeCredentialHolderParam`` if it has the next nonce following the sequential order.
- The revocation MUST fail if:

>     - The credential ID is not present in the registry.

Reviewer 3: Again consider adding the rejection codes to the standard

    - The credential status is not one of ``Active`` or ``NotActivated`` (see :ref:`CIS-4-functions-credentialStatus`).
    - The credential is not holder-revocable.
    - The signature was intended for a different contract.
    - The signature was intended for a different entrypoint.
    - The signature is expired.

>     - The signature can not be validated.

Reviewer 3: ```suggestion
    - The signature cannot be validated.
```

      The smart contract logic SHOULD practice its best efforts to ensure that only the holder can generate and authorize a revocation request with a valid signature.

.. _CIS-4-functions-revokeCredentialOther:

``revokeCredentialOther``
^^^^^^^^^^^^^^^^^^^^^^^^^^

Revoke a credential by a revocation authority request.
A revocation authority is any entity that holds a private key corresponding to the public key registered by the issuer.

> A revocation authority is authorized to revoke a credential by verifying the signature with the public key this the given identifier.

Reviewer: Something is off towards the end of this sentence.

Author: Fixed.

> A revocation authority is authorized to revoke a credential by verifying the signature with the public key this the given identifier.

Reviewer 3: ```suggestion
A revocation authority is authorized to revoke a credential by verifying the signature with the public key of the given identifier.
```

This entrypoint gives a general way of adding revocation rights to external entities.
It replaces the authorization checks conducted on the ``sender/invoker`` variable with signature verification.
In particular, it enables the issuer to provide a service for selected entities to revoke credentials without paying for revocation transactions.

Parameter
~~~~~~~~~

> It is serialized as :ref:`CIS-4-SignatureEd25519` (``signature``) and message data ``RevocationDataOther`` consisting of :ref:`CIS-4-CredentialID` (``credential_id``), metadata about the signature :ref:`CIS-4-SigningData` (``signing_data``), a reovocation public key :ref:`CIS-4-PublicKeyEd25519` , and an optional revocation reason (``reason``), serialized similarly to :ref:`CIS-4-functions-revokeCredentialIssuer`::

Reviewer: ```suggestion
It is serialized as :ref:`CIS-4-SignatureEd25519` (``signature``) and message data ``RevocationDataOther`` consisting of :ref:`CIS-4-CredentialID` (``credential_id``), metadata about the signature :ref:`CIS-4-SigningData` (``signing_data``), a revocation public key :ref:`CIS-4-PublicKeyEd25519` , and an optional revocation reason (``reason``), serialized similarly to :ref:`CIS-4-functions-revokeCredentialIssuer`::
```

  RevocationDataOther ::= (credential_id: CredentialID) (signing_data: SigningData) (revocation_key: PublicKeyEd25519) (reason: OptionReason)
  RevokeCredentialHolderParam ::= (signature: SignatureEd25519) (data : RevocationDataOther)

Requirements
~~~~~~~~~~~~

- If revoked successfully, the credential status MUST change to ``Revoked`` (see :ref:`CIS-4-functions-credentialStatus`).
- The ``RevokeCredentialOtherParam``'s ``signing_data`` MUST include a nonce to protect against replay attacks.
  The revocation authority's nonce is sequentially increased every time a revocation request is successfully executed.
  The function MUST only accept a ``RevokeCredentialOtherParam`` if it has the next nonce following the sequential order.
- The revocation MUST fail if:
    - The credential ID is not present in the registry.

>     - The revocation key in unknown.

Reviewer: Can we use the not present phrasing here as well?

Author: Done.

    - The credential status is not one of ``Active`` or ``NotActivated`` (see :ref:`CIS-4-functions-credentialStatus`).
    - The signature was intended for a different contract.
    - The signature was intended for a different entrypoint.
    - The signature is expired.
    - The signature can not be validated.
      The smart contract logic SHOULD practice its best efforts to ensure that only the revocation authority can generate and authorize a revocation request with a valid signature.

.. _CIS-4-functions-registerRevocationKeys:

``registerRevocationKeys``
^^^^^^^^^^^^^^^^^^^^^^^^^^

Register public keys that can be used by revocation authorities.

Parameter
~~~~~~~~~

It is serialized as First 2 bytes encode the length (``n``) the vector of kesy, followed by this many :ref:`CIS-4-PublicKeyEd25519` keys::

  RegisterPublicKeyParameters ::= (n: Byte²) (key: PublicKeyEd25519)ⁿ

Requirements
~~~~~~~~~~~~

- The revocation MUST fail if:

>     - The the transaction ``sender`` is not the issuer.

Reviewer 3: ```suggestion
    - The sender of the transaction is not the issuer.
```

    - Some of the keys are already registered.
- The smart contract MUST prevent resetting the nonce associated with a public key.
  For example, the contract logic could keep track of all keys seen by the contract and avoid reusing the same keys even after the keys was made unavailable by calling :ref:`CIS-4-functions-removeRevocationKeys`.

.. _CIS-4-functions-removeRevocationKeys:

``removeRevocationKeys``
^^^^^^^^^^^^^^^^^^^^^^^^

Make a list of public keys unavailable to revocation authorities.

Parameter
~~~~~~~~~

It is serialized as: First 2 bytes encode the length (``n``) the vector of keys, followed by this many :ref:`CIS-4-PublicKeyEd25519` keys::

  RegisterPublicKeyParameters ::= (n: Byte²) (key: PublicKeyEd25519)ⁿ

Requirements
~~~~~~~~~~~~

- The revocation MUST fail if:
    - The the transaction ``sender`` is not the issuer.

>     - Some of the keys are unknown.

Reviewer: Present in.. ;)

.. _CIS-4-functions-revocationKeys:

``revocationKeys``
^^^^^^^^^^^^^^^^^^

Query revocation keys.

Response
~~~~~~~~

The function output a list of available revocation keys.
Valid signatures with the corresponding private keys can be used to revoke any credential in the registry.

It is serialized as: First 2 bytes encode the length (``n``) the vector of keys, followed by this many :ref:`CIS-4-PublicKeyEd25519` keys::

  RegisterPublicKeyParameters ::= (n: Byte²) (key: PublicKeyEd25519)ⁿ

> Logged events

Author: Yes, this is a good point.
I think @abizjak was against it, but I don't remember why :)

> Logged events

Reviewer 2: You mean not to make them clash with tags from other standards?

While it's true you cannot make CIS-2 + CIS-4 supporting contract, it is hard to imagine such a use case.

Having to have all distinct tags from different standards is hopeless.

I don't see this being a problem and I don't want to change things yet again.

> Logged events

Reviewer 2: If this was a concern the only real solution would be to include the name of the standard in the event.

So we'd have events

```
struct Event {
   standard: Standard,
   standard_event: StandardDefinedEvent
}
```
where `Standard` would be names, so `CIS-0`, `CIS-1`, ...
and would be emitting these instead of requiring a globally unique u8 tag.

> Logged events

Reviewer 2: Or instead of the full `CIS-?` it could just be a u16 or u32 with the standard ID.

> Logged events

Reviewer 2: If you think there is a case where this will be a problem we can still change it now, it's just annoying since there are many components.

> Logged events

Reviewer 2: There is also CIS3 which has this 250 event for nonce. It is conceivable to sponsor issuance perhaps so we need to then not use 250.

I don't want to spend more time revising serialization so we'll then just use 249.. codes for now.

But any future standard we must revise this, because it is not tenable to cram all possible event types into a u8 globally.

> Logged events

Author: I like the idea of using standard ids as tags in future CIS  standards.

> Logged events

Author: I fixed the event tags. Now they start with 249 counting down.

> Logged events

Reviewer 3: yes, we should have a better solution in the future

-------------

The events defined by this specification are serialized using one byte to discriminate the different events.
A custom event SHOULD NOT have a first byte colliding with any of the events defined by this specification.

.. _CIS-4-register-credential-transfer:

``RegisterCredentialEvent``
^^^^^^^^^^^^^^^^^^^^^^^^^^^

A ``RegisterCredentialEvent`` event MUST be logged when a new credential is registered with :ref:`CIS-4-functions-registerCredential`.

> The ``RegisterCredentialEvent`` event is serialized as: a first a byte with the value of 255, followed by :ref:`CIS-4-CredentialID` (``crednetial_id``), a public key of the holder :ref:`CIS-4-PublicKeyEd25519` (``holder_id``), a reference to the credential schema :ref:`CIS-4-SchemaRef` (``schema_ref``), a credential type :ref:`CIS-4-CredentialType` (``credential_type``) ::

Reviewer: ```suggestion
The ``RegisterCredentialEvent`` event is serialized as: first a byte with the value of 255, followed by :ref:`CIS-4-CredentialID` (``crednetial_id``), a public key of the holder :ref:`CIS-4-PublicKeyEd25519` (``holder_id``), a reference to the credential schema :ref:`CIS-4-SchemaRef` (``schema_ref``), a credential type :ref:`CIS-4-CredentialType` (``credential_type``) ::
```

The ``RevokeCredentialEvent`` event is serialized as: a first a byte with the value of 254, followed by :ref:`CIS-4-CredentialID` (``crednetial_id``), a public key of the holder :ref:`CIS-4-PublicKeyEd25519` (``holder_id``), a ``revoker``, and an optional revocation reason (``reason``), serialized similarly to :ref:`CIS-4-functions-revokeCredentialIssuer`; ``revoker`` is serialized as 1 byte to indicate who sent the revocation request ( 0 - issuer, 1 - holder, 2 -revocation authority); if the first byte is 2, then it is followed by a public key :ref:`CIS-4-PublicKeyEd25519` of the revoker::

  Revoker ::= (0: Byte)                         // Issuer
            | (1: Byte)                         // Holder
            | (2: Byte) (key: PublicKeyEd25519) // Other
  RevokeCredentialEvent ::= (254: Byte) (credential_id: CredentialID) (holder_id: PublicKeyEd25519) (revoker: Revoker) (reason: OptionReason)

``IssuerMetadata``
^^^^^^^^^^^^^^^^^^

A ``IssuerMetadata`` event MUST be logged when updating the credential metadata.

> This also applies to contract initialization.

Reviewer 2: What is your comment here? That we don't want to necessarily emit the metadata event in initialization?

If we require it then it is simple to know the metadata from events by just looking at events. THere is no special casing needed. Or is your comment about something else?

> This also applies to contract initialization.

Author: I agree with @abizjak, I think it is useful to log the event in initialization. @limemloh what do you think? Should it be phrased differently?

> This also applies to contract initialization.

Reviewer 3: Right, this was not very clear. The event is good to have, but I don't think we should have a requirement for it to be emitted during contract initialization.

It consists of a URL for the location of the metadata for the issuer with an optional SHA256 checksum of the content.

The ``IssuerMetadata`` event is serialized as: first a byte with the value of 253, followed by :ref:`CIS-2-MetadataUrl` (``metadata``)::

  IssuerMetadata ::= (253: Byte) (metadata: MetadataUrl)

``CredentialMetadataEvent``
^^^^^^^^^^^^^^^^^^^^^^^^^^^

A ``CredentialMetadataEvent`` event MUST be logged when updating the credential metadata.
It consist of a credential ID and a URL for the location of the metadata for this credential with an optional SHA256 checksum of the content.

The ``TokenMetadataEvent`` event is serialized as: first a byte with the value of 252, followed by the token ID :ref:`CIS-2-TokenID` (``id``), and then a :ref:`CIS-2-MetadataUrl` (``metadata``)::

>   TokenMetadataEvent ::= (252: Byte) (id: CredentialID) (metadata: MetadataUrl)

Author: Fixed.

  Nonce ::= (nonce: Byte⁸)

.. _CIS-4-PublicKeyEd25519:

``PublicKeyEd25519``
^^^^^^^^^^^^^^^^^^^^

A public key is represented as a 32-byte array.

It is serialized as 32 bytes::

  PublicKeyEd25519 ::= (key: Byte³²)

.. _CIS-4-SignatureEd25519:

``SignatureEd25519``
^^^^^^^^^^^^^^^^^^^^

Signature for an Ed25519 message.

It is serialized as 64 bytes::

  SignatureEd25519 ::= (signature: Byte⁶⁴)

.. _CIS-4-SigningData:

``SigningData``
^^^^^^^^^^^^^^^

Signing data contains a metadata for the signature that is used to check whether the signed message is designated for the right contract and entrypoint, and it is not expired.

It is serialized as :ref:`CIS-4-ContractAddress` (``contract_address``), :ref:`CIS-4-EntrypointName` (``entrypoint``), :ref:`CIS-4-Nonce` (``nonce``), and :ref:`CIS-4-Timestamp` (``timestamp``)::

  SigningData ::= (contract_address: ContractAddress) (entrypoint: EntrypointName) (nonce: Nonce) (timestamp: Timestamp)

.. _CIS-4-SchemaRef:

``SchemaRef``
^^^^^^^^^^^^^

A URL of the credential schema.

Serialized in the same way as :ref:`CIS-2 MetadataUrl<CIS-2-MetadataUrl>`.

.. _CIS-4-CredentialType:

``CredentialType``
^^^^^^^^^^^^^^^^^^

> Is a short ASCII string (up to 256 characters) describing the credential type that is used to identify which schema the credential is based on.

Author: Changed to UTF-8.

  OptionalTimestamp ::= (0: Byte)
                      | (1: Byte) (timestamp: Timestamp)
  CredentialInfo ::= (holder_id: CredentialHolderId) (holder_revocable: Bool) (commitment: Commitment) (valid_from: Timestamp)
                     (valid_until: OptionTimestamp) (credential_type: CredentialType) (metadata_url: MetadataUrl)

.. note::
  The timestamp ``valid_until`` is optional; if it is not included (indicated by the 0 value), then the credential never expires.

.. _CIS-4-CredentialStatus:

``CredentialStatus``
^^^^^^^^^^^^^^^^^^^^

The status of a verifiable credential.

> It is serialized as 1 byte where ``0`` correponds to the status ``Active``, ``1`` corresponds to  ``Revoked``, ``2`` corresponds to  ``Expired``, ``3`` corresponds to ``NotActivated``::

Author: I explained the meaning of statuses in the requirements for the `credentialStatus` entrypoint. I'll  add reference here.

^^^^^^^^^^

Query the issuer's account address.

Response
~~~~~~~~

The function output is the issuer's account address.
It is serialized as :ref:`CIS-4-AccountAddress`.

.. _CIS-4-functions-issuerMetadata:

``issuerMetadata``
^^^^^^^^^^^^^^^^^^

Query the current token metadata URLs for a list of token IDs.

Response
~~~~~~~~

The function output is the issuer's metadata URL.

It is serialized as :ref:`CIS-2-MetadataUrl`.

.. _CIS-4-functions-registerCredential:

``registerCredential``
^^^^^^^^^^^^^^^^^^^^^^

Register a credential with the given ID.

Parameter
~~~~~~~~~

The parameter is the credential ID and credential information that is used to create an entry in the registry.

It is serialized as :ref:`CIS-4-CredentialHolderId` (``credential_id``) followed by :ref:`CIS-4-CredentialInfo` (``credential_info``)::

  RegisterCredentialParameter ::= (credential_id: CredentialHolderId) (credential_info: CredentialInfo)

Requirements
~~~~~~~~~~~~

- The credential registration request MUST fail if the credential ID is already present in the registry.
- After successful registration:
    - Querying the credential by its ID with :ref:`CIS-4-functions-credentialEntry` MUST succeed.

>     - Querying the credential status by ID with :ref:`CIS-4-functions-credentialStatus` MUST succeed and MUST NOT return ``Revoked`` (See the possible values for the status in :ref:`CIS-4-CredentialStatus`).

Author: My reasoning was that revocation must always be explicit generating a revocation event. In this case, it's clear that if there is no revocation event in the log, the credential is not revoked. However, it was before we introduced the possibility of restoring revoked credentials. With the restoration functionality, a revocation event does not mean that the credential stays revoked. Indeed, it seems like it's not necessary to specify this.

.. _CIS-4-functions-revokeCredentialIssuer:

``revokeCredentialIssuer``
^^^^^^^^^^^^^^^^^^^^^^^^^^

Revoke a credential by the issuer's request.
The issuer is authorized to revoke a credential if the transaction sender's address is the same as the return value of :ref:`CIS-4-functions-issuer`.

Parameter
~~~~~~~~~

The parameter is the credential ID :ref:`CIS-4-CredentialHolderId` and optional string indicating the revocation reason.

It is serialized as :ref:`CIS-4-CredentialHolderId` followed by 1 byte to indicate whether a reason is included.
If its value is 0, then no reason string is present, if the value is 1 then the bytes corresponding to the reason string follow::

  OptionalReason ::= (0: Byte)
                   | (1: Byte) (n: Byte) (reason_string: Byteⁿ)
  RevokeCredentialIssuerParam ::= (credential_id: CredentialHolderId) (reason: OptionReason)

> .. TODO: what kind of characters are allowed? ASCII, Unicode?

Author: Done.

Requirements
~~~~~~~~~~~~

- If revoked successfully, the credential status MUST change to ``Revoked`` (see :ref:`CIS-4-functions-credentialStatus`).
- The revocation MUST fail if:
    - The sender of the transaction is not the issuer.
    - The credential ID is not present in the registry.
    - The credential status is not one of ``Active`` or ``NotActivated`` (see :ref:`CIS-4-functions-credentialStatus`).

.. _CIS-4-functions-revokeCredentialHolder:

``revokeCredentialHolder``
^^^^^^^^^^^^^^^^^^^^^^^^^^

Revoke a credential by the holders's request.

The holder is authorized to revoke a credential by verifying the signature with the holder's public key.

> It replaces the authorization based on checking the transaction sender address with signature verification.

Reviewer 2: Why are we not specifying the exact message to be signed, with the domain separation string included?

Author: Yes, good point. I'll add the message format.

Parameter
~~~~~~~~~

It is serialized as :ref:`CIS-4-SignatureEd25519` (``signature``) and message data ``RevocationDataHolder`` consisting of :ref:`CIS-4-CredentialHolderId` (``credential_id``), metadata about the signature :ref:`CIS-4-SigningData` (``signing_data``), and an optional revocation reason (``reason``), serialized similarly to :ref:`CIS-4-functions-revokeCredentialIssuer`::

  RevocationDataHolder ::= (credential_id: CredentialHolderId) (signing_data: SigningData) (reason: OptionReason)
  RevokeCredentialHolderParam ::= (signature: SignatureEd25519) (data : RevocationDataHolder)

Requirements
~~~~~~~~~~~~

- If revoked successfully, the credential status MUST change to ``Revoked`` (see :ref:`CIS-4-functions-credentialStatus`).
- The ``RevokeCredentialHolderParam``'s ``signing_data`` MUST include a nonce to protect against replay attacks.
  The holders's nonce is sequentially increased every time a revocation request is successfully executed.
  The function MUST only accept a ``RevokeCredentialHolderParam`` if it has the next nonce following the sequential order.

> - The revocation MUST fail if:

Reviewer 2: Make it clear that the list is an "or"., i.e., `must fail if any of the following conditions are met`.

Author: Done.

Requirements
~~~~~~~~~~~~

- The revocation MUST fail if:
    - The sender of the transaction is not the issuer.
    - Some of the keys are already registered.
- The smart contract MUST prevent resetting the nonce associated with a public key.
  For example, the contract logic could keep track of all keys seen by the contract and avoid reusing the same keys even after the keys were made unavailable by calling :ref:`CIS-4-functions-removeRevocationKeys`.

.. _CIS-4-functions-removeRevocationKeys:

``removeRevocationKeys``
^^^^^^^^^^^^^^^^^^^^^^^^

Make a list of public keys unavailable to revocation authorities.

Parameter
~~~~~~~~~

It is serialized as: First 2 bytes encode the length (``n``) the vector of keys, followed by this many :ref:`CIS-4-PublicKeyEd25519` keys::

  RegisterPublicKeyParameters ::= (n: Byte²) (key: PublicKeyEd25519)ⁿ

Requirements
~~~~~~~~~~~~

- The revocation MUST fail if:
    - The sender of the transaction is not the issuer.
    - Some of the keys are not present in the registry.

.. _CIS-4-functions-revocationKeys:

``revocationKeys``
^^^^^^^^^^^^^^^^^^

Query revocation keys.

Response
~~~~~~~~

> The function output a list of available revocation keys.

Reviewer 2: ```suggestion
The function outputs a list of available revocation keys.
```

Author: Fixed.

To associate a hash with a URL the JSON value is an object:

.. list-table:: URL JSON Object
  :header-rows: 1

  * - Property
    - JSON value type [JSON-Schema]
    - Description
  * - ``url``
    - string (:rfc:`3986`) [``uri-reference``]
    - A URL.
  * - ``hash`` (optional)
    - string
    - A SHA256 hash of the URL content encoded as a hex string.

Example issuer metadata
^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: json

  {
    "name": "Concordium",
    "icon" : {
      "url":  "https://concordium.com/wp-content/uploads/2022/07/Concordium-1.png",
      "hash": "1c74f7eb1b3343a5834e60e9a8fce277f2c7553112accd42e63fae7a09e0caf8"
      }
    "description": "A public-layer 1, science-backed blockchain",
    "url": "https://concordium.com"
  }

Credential metadata JSON
------------------------

The credential metadata is stored off-chain and MUST be a JSON (:rfc:`8259`) file.

.. list-table:: Credential metadata JSON Object
  :header-rows: 1

  * - Property
    - JSON value type [JSON-Schema]
    - Description
  * - ``title``
    - string
    - The name to display for the credential.

>   * - ``logo``

Reviewer 4: Do we want to specify formats (sizes) in the standard document? Or do we want to do that elsewhere?

Author: Good question. CIS-2 does not specify the sizes. I think it makes sense to leave them out. CIS standards are about the API, additional details might be specified elsewhere.

@orhoj is there a document that specifies the image sizes for CIS-2 complaint NFTs?

Author: Let's add a recommendation about the image sizes to the standard. This recommendation might be revised later depending on the final version of the wallet's UI.

Author: Done. I added initial values and a TODO to check before we finalize the standard.

    - URL JSON object
    - An image URL for displaying the credential.
  * - ``background_color``
    - string
    - A hex code of the background color for displaying the credential.
  * - ``image`` (optional)
    - URL JSON object
    - A background image URL for displaying the credential.

Where URL JSON object the same as in :ref:`CIS-4-events-IssuerMetadata`.

Optionally a SHA256 hash of the JSON file can be logged with the :ref:`CIS-4-events-CredentialMetadataEvent` event for checking integrity.

Example credential metadata
^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: json

 {
    "title": "Concordium Employment",
    "logo" : {
      "url":  "https://concordium.com/wp-content/uploads/2022/07/Concordium-1.png",
      "hash": "1c74f7eb1b3343a5834e60e9a8fce277f2c7553112accd42e63fae7a09e0caf8"
      }
    "background_color": "#000000",
    "image": {

>       "url": "https://concordium.com/employment/vc-background.png",

Reviewer 4: There is a hash for the `logo` but not for the `image`. Should it be the same?

Author: `hash` is optional, so I wanted to give an example with and without `hash`. I can add a comment explaining this.

Requirements
~~~~~~~~~~~~

- The revocation MUST fail if some of the keys are not present in the registry.

.. _CIS-4-functions-revocationKeys:

``revocationKeys``
^^^^^^^^^^^^^^^^^^

Query revocation keys.

Response
~~~~~~~~

The function outputs a list of available revocation keys.
Valid signatures with the corresponding private keys can be used to revoke any credential in the registry.

It is serialized as: First 2 bytes encode the length (``n``) of the vector of keys, followed by this many :ref:`CIS-4-PublicKeyEd25519` keys::

  RegisterPublicKeyParameters ::= (n: Byte²) (key: PublicKeyEd25519)ⁿ

Logged events
-------------

The events defined by this specification are serialized using one byte to discriminate the different events.
A custom event SHOULD NOT have a first byte colliding with any of the events defined by this specification.

.. _CIS-4-register-credential-transfer:

``RegisterCredentialEvent``
^^^^^^^^^^^^^^^^^^^^^^^^^^^

A ``RegisterCredentialEvent`` event MUST be logged when a new credential is registered.
The event records the credential identifier, the credential type, and the corresponding schema reference.

> The ``RegisterCredentialEvent`` event is serialized as: first a byte with the value of 255, followed by :ref:`CIS-4-CredentialHolderId` (``crednetial_id``), a reference to the credential schema :ref:`CIS-4-SchemaRef` (``schema_ref``), a credential type :ref:`CIS-4-CredentialType` (``credential_type``) ::

Reviewer 3: ```suggestion
The ``RegisterCredentialEvent`` event is serialized as: first a byte with the value of 249, followed by :ref:`CIS-4-CredentialHolderId` (``crednetial_id``), a reference to the credential schema :ref:`CIS-4-SchemaRef` (``schema_ref``), a credential type :ref:`CIS-4-CredentialType` (``credential_type``) ::
```

  CredentialEventData ::= (credential_id: CredentialHolderId) (schema_ref: SchemaRef) (credential_type: CredentialType)
  RegisterCredentialEvent ::= (249: Byte) (data: CredentialEventData)

``RevokeCredentialEvent``
^^^^^^^^^^^^^^^^^^^^^^^^^

A ``RevokeCredentialEvent`` event MUST be logged when a credential is revoked.
The event records the credential identifier, who requested the revocation (the holder, the issuer or a revocation authority), and an optional string with a short comment on the revocation reason.

> The ``RevokeCredentialEvent`` event is serialized as: first a byte with the value of 254, followed by :ref:`CIS-4-CredentialHolderId` (``crednetial_id``), a ``revoker``, and an optional revocation reason (``reason``), serialized similarly to :ref:`CIS-4-functions-revokeCredentialIssuer`; ``revoker`` is serialized as 1 byte to indicate who sent the revocation request ( 0 - issuer, 1 - holder, 2 -revocation authority); if the first byte is 2, then it is followed by a public key :ref:`CIS-4-PublicKeyEd25519` of the revoker::

Reviewer 3: ```suggestion
The ``RevokeCredentialEvent`` event is serialized as: first a byte with the value of 248, followed by :ref:`CIS-4-CredentialHolderId` (``crednetial_id``), a ``revoker``, and an optional revocation reason (``reason``), serialized similarly to :ref:`CIS-4-functions-revokeCredentialIssuer`; ``revoker`` is serialized as 1 byte to indicate who sent the revocation request ( 0 - issuer, 1 - holder, 2 -revocation authority); if the first byte is 2, then it is followed by a public key :ref:`CIS-4-PublicKeyEd25519` of the revoker::
```

  Revoker ::= (0: Byte)                         // Issuer
            | (1: Byte)                         // Holder
            | (2: Byte) (key: PublicKeyEd25519) // Other
  RevokeCredentialEvent ::= (248: Byte) (credential_id: CredentialHolderId) (revoker: Revoker) (reason: OptionalReason)

.. _CIS-4-events-IssuerMetadata:

``IssuerMetadata``
^^^^^^^^^^^^^^^^^^

A ``IssuerMetadata`` event MUST be logged when setting the metadata url of the issuer.
It consists of a URL for the location of the metadata for the issuer with an optional SHA256 checksum of the content.

> The ``IssuerMetadata`` event is serialized as: first a byte with the value of 253, followed by :ref:`CIS-2-MetadataUrl` (``metadata``)::

Reviewer 3: ```suggestion
The ``IssuerMetadata`` event is serialized as: first a byte with the value of 247, followed by :ref:`CIS-2-MetadataUrl` (``metadata``)::
```

    - The credential status is not one of ``Active`` or ``NotActivated`` (see :ref:`CIS-4-functions-credentialStatus`).
    - The signature was intended for a different contract.
    - The signature was intended for a different entrypoint.
    - The signature is expired.
    - The signature can not be validated.
      The smart contract logic SHOULD practice its best efforts to ensure that only the revocation authority can generate and authorize a revocation request with a valid signature.

.. _CIS-4-functions-registerRevocationKeys:

``registerRevocationKeys``
^^^^^^^^^^^^^^^^^^^^^^^^^^

Register public keys that can be used by revocation authorities.

.. note::
  This standard does not specify how the issuer is authenticated.
  Implementations can use various mechanisms.
  For example, the transaction sender's address is checked against the issuer's account address stored in the contract.
  Another option is to use ``auxiliary_data`` to implement signature-based authentication mechanism.

Parameter
~~~~~~~~~

It is serialized as First 2 bytes encode the length (``n``) of the vector of keys, followed by this many :ref:`CIS-4-PublicKeyEd25519` keys::

>   RegisterPublicKeyParameters ::= (n: Byte²) (key: PublicKeyEd25519)ⁿ

Reviewer 5: In the `credential-registry` contract example, this also includes auxiliary data (I see you included it in the note, so maybe you just forgot to add it here as well?)
```suggestion
  AuxData ::= (n: Byte²) (data: Byteⁿ)
  RegisterPublicKeyParameters ::= (n: Byte²) (key: PublicKeyEd25519)ⁿ (auxiliary_data: AuxData)
```

Author: Yep, nice catch! Thanks!

Requirements
~~~~~~~~~~~~

- The revocation MUST fail if some of the keys are already registered.
- The smart contract MUST prevent resetting the nonce associated with a public key.
  For example, the contract logic could keep track of all keys seen by the contract and avoid reusing the same keys even after the keys were made unavailable by calling :ref:`CIS-4-functions-removeRevocationKeys`.

.. _CIS-4-functions-removeRevocationKeys:

``removeRevocationKeys``
^^^^^^^^^^^^^^^^^^^^^^^^

Make a list of public keys unavailable to revocation authorities.

.. note::
  This standard does not specify how the issuer is authenticated.
  Implementations can use various mechanisms.
  For example, the transaction sender's address is checked against the issuer's account address stored in the contract.
  Another option is to use ``auxiliary_data`` to implement signature-based authentication mechanism.

Parameter
~~~~~~~~~

It is serialized as: First 2 bytes encode the length (``n``) of the vector of keys, followed by this many :ref:`CIS-4-PublicKeyEd25519` keys::

>   RegisterPublicKeyParameters ::= (n: Byte²) (key: PublicKeyEd25519)ⁿ

Reviewer 5: And here:
```suggestion
  AuxData ::= (n: Byte²) (data: Byteⁿ)
  RegisterPublicKeyParameters ::= (n: Byte²) (key: PublicKeyEd25519)ⁿ (auxiliary_data: AuxData)
```

Author: Indeed, thanks!

.. TODO: refer to the Concordium VC Data Model documentation, once we have it.

Specification
=============

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in :rfc:`2119`.

General types and serialization
-------------------------------

.. _CIS-4-Bool:

``Bool``
^^^^^^^^

A boolean is serialized as a byte with value 0 for false and 1 for true::

  Bool ::= (0: Byte) // False
         | (1: Byte) // True

.. _CIS-4-CredentialHolderId:

``CredentialHolderId``
^^^^^^^^^^^^^^^^^^^^^^

A credential identifier is a holder's public key.
It is expected that for each issued credential a unique public key is generated.
The credential identifier also identifies the credential holder and can be used to prove ownership of the credential by the holder.

It is serialized as :ref:`CIS-4-PublicKeyEd25519`.

.. _CIS-4-MetadataUrl:

``MetadataUrl``
^^^^^^^^^^^^^^^

A URL and optional checksum for metadata stored outside of this contract.

> It is serialized as: 2 bytes for the length of the metadata url (``n``) and then this many bytes for the url to the metadata (``url``) followed by an optional checksum.

Reviewer 6: ```suggestion
It is serialized as: 2 bytes for the length of the metadata URL (``n``) and then this many bytes for the URL to the metadata (``url``) followed by an optional checksum.
```

Reviewer 7: Do you want to mention the endianness of `n`?

Author: Yes, this is a good point! I wonder why it's not mentioned in CIS-2, where I copied the definition from. And what is the endianness we actually use when serializing/deserializing the size?

Author: Clarified with @abizjak, it's little endian. I think I can write a general remark about how the size is serialized, because it's used in several places.

The checksum is serialized by 1 byte to indicate whether a hash of the metadata is included.

> If its value is 0, then there is no hash, if the value is 1 then 32 bytes for a SHA256 hash (``hash``) follows::

Reviewer 6: ```suggestion
If its value is 0, then there is no hash; if the value is 1, then 32 bytes for a SHA256 hash (``hash``) follows::
```

Response
~~~~~~~~

The function output is the issuer's public key.
It is serialized as :ref:`CIS-4-PublicKeyEd25519`.

.. _CIS-4-functions-registryMetadata:

``registryMetadata``
^^^^^^^^^^^^^^^^^^^^

Query the registry's metadata.

Response
~~~~~~~~

The function output is the issuer's metadata URL, the credential type and schema for the credentials stored in the registry.

It is serialized as the issuer's :ref:`CIS-4-MetadataUrl` (``issuer_metadata``) followed by the credential type of the registry :ref:`CIS-4-CredentialType` (``credential_type``) and the corresponding credential JSON schema reference :ref:`CIS-4-SchemaRef` (``credential_schema``)::

  MetadataResponse := (issuer_metadata: MetadataUrl) (credential_type: CredentialType) (credential_schema: SchemaRef)

.. _CIS-4-functions-registerCredential:

``registerCredential``
^^^^^^^^^^^^^^^^^^^^^^

Register public data for a new credential.

.. note::
  This standard does not specify how the issuer is authenticated.
  Implementations can use various mechanisms.
  For example, the transaction sender's address is checked against the issuer's account address stored in the contract.

>   Another option is to use ``auxiliary_data`` to implement signature-based authentication mechanism.

Reviewer 6: ```suggestion
  Another option is to use ``auxiliary_data`` to implement a signature-based authentication mechanism.
```

Parameter
~~~~~~~~~

The parameter is credential information that is used to create an entry in the registry.

> It is serialized as :ref:`CIS-4-CredentialInfo` (``credential_info``) followed by auxiliary data, which is serialized as 2 bytes to encode the length (``n``) of the vector of keys, followed by this many bytes of data::

Reviewer 6: ```suggestion
It is serialized as :ref:`CIS-4-CredentialInfo` (``credential_info``), followed by auxiliary data, which is serialized as 2 bytes to encode the length (``n``) of the vector of keys, followed by this many bytes of data::
```

  AuxData ::= (n: Byte²) (data: Byteⁿ)
  RegisterCredentialParameter ::= (credential_info: CredentialInfo) (auxiliary_data: AuxData)

Requirements
~~~~~~~~~~~~

- The credential registration request MUST fail if the credential ID is already present in the registry.
- After successful registration, querying the credential by its ID with :ref:`CIS-4-functions-credentialEntry` MUST succeed.

.. _CIS-4-functions-revokeCredentialIssuer:

``revokeCredentialIssuer``
^^^^^^^^^^^^^^^^^^^^^^^^^^

Revoke a credential by the issuer's request.

.. note::
  This standard does not specify how the issuer is authenticated.
  Implementations can use various mechanisms.
  For example, the transaction sender's address is checked against the issuer's account address stored in the contract.

>   Another option is to use ``auxiliary_data`` to implement signature-based authentication mechanism.

Reviewer 6: ```suggestion
  Another option is to use ``auxiliary_data`` to implement a signature-based authentication mechanism.
```

Parameter
~~~~~~~~~

The parameter is the credential ID :ref:`CIS-4-CredentialHolderId` and an optional string in the UTF-8 encoding that indicates the revocation reason.

It is serialized as :ref:`CIS-4-CredentialHolderId` followed by 1 byte to indicate whether a reason is included.

> If its value is 0, then no reason string is present, if the value is 1 then the bytes corresponding to the reason string follow.

Reviewer 6: ```suggestion
If its value is 0, then no reason string is present; if the value is 1, then the bytes corresponding to the reason string follow.
```

Requirements
~~~~~~~~~~~~

- If revoked successfully, the credential status MUST change to ``Revoked`` (see :ref:`CIS-4-functions-credentialStatus`).
- The message to be signed MUST be produced in the following way:
    - Start with the bytes of the domain separation string ``WEB3ID:REVOKE``.
    - Append ``RevocationDataHolder`` bytes from the input parameter.
- The ``RevokeCredentialHolderParam``'s ``signing_data`` MUST include a nonce to protect against replay attacks.
  The holders's nonce is sequentially increased every time a revocation request is successfully executed.
  The function MUST only accept a ``RevokeCredentialHolderParam`` if it has the next nonce following the sequential order.
- The revocation MUST fail if any of the following conditions are met:
    - The credential ID is not present in the registry.
    - The credential status is not one of ``Active`` or ``NotActivated`` (see :ref:`CIS-4-functions-credentialStatus`).
    - The credential is not holder-revocable.
    - The signature was intended for a different contract.
    - The signature was intended for a different entrypoint.
    - The signature is expired.
    - The signature cannot be validated.

>       The smart contract logic SHOULD practice its best efforts to ensure that only the holder can generate and authorize a revocation request with a valid signature.

Author: Done, and the one below, too.

- If revoked successfully, the credential status MUST change to ``Revoked`` (see :ref:`CIS-4-functions-credentialStatus`).
- The message to be signed MUST be produced in the following way:
    - Start with the bytes of the domain separation string ``WEB3ID:REVOKE``.
    - Append ``RevocationDataOther`` bytes from the input parameter.
- The ``RevokeCredentialOtherParam``'s ``signing_data`` MUST include a nonce to protect against replay attacks.
  The revocation authority's nonce is sequentially increased every time a revocation request is successfully executed.
  The function MUST only accept a ``RevokeCredentialOtherParam`` if it has the next nonce following the sequential order.
- The revocation MUST fail if any of the following conditions are met:
    - The credential ID is not present in the registry.
    - The revocation key in not present in the registry.
    - The credential status is not one of ``Active`` or ``NotActivated`` (see :ref:`CIS-4-functions-credentialStatus`).
    - The signature was intended for a different contract.
    - The signature was intended for a different entrypoint.
    - The signature is expired.
    - The signature can not be validated.

>       The smart contract logic SHOULD practice its best efforts to ensure that only the revocation authority can generate and authorize a revocation request with a valid signature.

Reviewer 6: Same as above.

The revocation keys are followed by auxiliary data, which is serialized as 2 bytes to encode the length (``m``) of the data vector, followed by this many bytes of data::

  AuxData ::= (m: Byte²) (data: Byteᵐ)
  RegisterPublicKeyParameters ::= (n: Byte²) (key: PublicKeyEd25519)ⁿ (auxiliary_data: AuxData)

Requirements
~~~~~~~~~~~~

- The revocation MUST fail if some of the keys are already registered.
- The smart contract MUST prevent resetting the nonce associated with a public key.
  For example, the contract logic could keep track of all keys seen by the contract and avoid reusing the same keys even after the keys were made unavailable by calling :ref:`CIS-4-functions-removeRevocationKeys`.

.. _CIS-4-functions-removeRevocationKeys:

``removeRevocationKeys``
^^^^^^^^^^^^^^^^^^^^^^^^

Make a list of public keys unavailable to revocation authorities.

.. note::
  This standard does not specify how the issuer is authenticated.
  Implementations can use various mechanisms.
  For example, the transaction sender's address is checked against the issuer's account address stored in the contract.

>   Another option is to use ``auxiliary_data`` to implement signature-based authentication mechanism.

Reviewer 6: ```suggestion
  Another option is to use ``auxiliary_data`` to implement a signature-based authentication mechanism.
```

Logged events
-------------

The events defined by this specification are serialized using one byte to discriminate the different events.
A custom event SHOULD NOT have a first byte colliding with any of the events defined by this specification.

.. _CIS-4-register-credential-transfer:

``RegisterCredentialEvent``
^^^^^^^^^^^^^^^^^^^^^^^^^^^

A ``RegisterCredentialEvent`` event MUST be logged when a new credential is registered.
The event records the credential identifier, the credential type, and the corresponding schema reference.

The ``RegisterCredentialEvent`` event is serialized as: first a byte with the value of 249, followed by :ref:`CIS-4-CredentialHolderId` (``crednetial_id``), a reference to the credential schema :ref:`CIS-4-SchemaRef` (``schema_ref``), a credential type :ref:`CIS-4-CredentialType` (``credential_type``) ::

  CredentialEventData ::= (credential_id: CredentialHolderId) (schema_ref: SchemaRef) (credential_type: CredentialType)
  RegisterCredentialEvent ::= (249: Byte) (data: CredentialEventData)

``RevokeCredentialEvent``
^^^^^^^^^^^^^^^^^^^^^^^^^

A ``RevokeCredentialEvent`` event MUST be logged when a credential is revoked.

> The event records the credential identifier, who requested the revocation (the holder, the issuer or a revocation authority), and an optional string with a short comment on the revocation reason.

Reviewer 6: ```suggestion
The event records the credential identifier who requested the revocation (the holder, the issuer, or a revocation authority), and an optional string with a short comment on the revocation reason.
```

> The ``RevokeCredentialEvent`` event is serialized as: first a byte with the value of 248, followed by :ref:`CIS-4-CredentialHolderId` (``crednetial_id``), a ``revoker``, and an optional revocation reason (``reason``), serialized similarly to :ref:`CIS-4-functions-revokeCredentialIssuer`; ``revoker`` is serialized as 1 byte to indicate who sent the revocation request ( 0 - issuer, 1 - holder, 2 -revocation authority); if the first byte is 2, then it is followed by a public key :ref:`CIS-4-PublicKeyEd25519` of the revoker::

Reviewer 6: ```suggestion
The ``RevokeCredentialEvent`` event is serialized as: first a byte with the value of 248, followed by :ref:`CIS-4-CredentialHolderId` (``crednetial_id``), a ``revoker``, and an optional revocation reason (``reason``), serialized similarly to :ref:`CIS-4-functions-revokeCredentialIssuer`. ``revoker`` is serialized as 1 byte to indicate who sent the revocation request ( 0 - issuer, 1 - holder, 2 -revocation authority); if the first byte is 2, then it is followed by a public key :ref:`CIS-4-PublicKeyEd25519` of the revoker::
```

The ``CredentialMetadataEvent`` event is serialized as: first a byte with the value of 246, followed by :ref:`CIS-4-CredentialHolderID` (``id``), and then a :ref:`CIS-4-MetadataUrl` (``metadata``)::

  CredentialMetadataEvent ::= (246: Byte) (id: CredentialHolderId) (metadata: MetadataUrl)

``CredentialSchemaRefEvent``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

A ``CredentialSchemaRefEvent`` event MUST be logged when updating the credential schema reference for a credential type.
It consist of a credential type and a URL for the location of the schema for this credential with an optional SHA256 checksum of the content.

The ``CredentialSchemaRefEvent`` event is serialized as: first a byte with the value of 245, followed by :ref:`CIS-4-CredentialType` (``type``), and then a :ref:`CIS-4-SchemaRef` (``schema_ref``)::

  CredentialSchemaRefEvent ::= (245: Byte) (type: CredentialType) (schema_ref: SchemaRef)

``RevocationKeyEvent``
^^^^^^^^^^^^^^^^^^^^^^

A ``RevocationKeyEvent`` event MUST be logged when registering a new or removing an existing revocation key.

> It consist of the key and the action performed with the key (registration or removal).

Reviewer 6: ```suggestion
It consists of the key and the action performed with the key (registration or removal).
```

^^^^^^^^^

An unsigned 64-bit integer number that increases sequentially to protect against replay attacks.

It is serialized as 8 bytes in little endian::

  Nonce ::= (nonce: Byte⁸)

.. _CIS-4-PublicKeyEd25519:

``PublicKeyEd25519``
^^^^^^^^^^^^^^^^^^^^

A public key is represented as a 32-byte array.

It is serialized as 32 bytes::

  PublicKeyEd25519 ::= (key: Byte³²)

.. _CIS-4-SignatureEd25519:

``SignatureEd25519``
^^^^^^^^^^^^^^^^^^^^

Signature for an Ed25519 message.

It is serialized as 64 bytes::

  SignatureEd25519 ::= (signature: Byte⁶⁴)

.. _CIS-4-SigningData:

``SigningData``
^^^^^^^^^^^^^^^

Signing data contains a metadata for the signature that is used to check whether the signed message is designated for the right contract and entrypoint, and it is not expired.

It is serialized as :ref:`CIS-4-ContractAddress` (``contract_address``), :ref:`CIS-4-EntrypointName` (``entrypoint``), :ref:`CIS-4-Nonce` (``nonce``), and :ref:`CIS-4-Timestamp` (``timestamp``)::

  SigningData ::= (contract_address: ContractAddress) (entrypoint: EntrypointName) (nonce: Nonce) (timestamp: Timestamp)

.. _CIS-4-SchemaRef:

``SchemaRef``
^^^^^^^^^^^^^

A URL of the credential schema.

> Serialized in the same way as :ref:`CIS-2 MetadataUrl<CIS-2-MetadataUrl>`.

Reviewer 5: Do we maybe want to link to the CIS4 metadataURL serialization instead? It seems odd to me that we duplicate the serialized format and then link to the CIS2 version here. Linking to the CIS4 reference also helps the user stay in the document instead of jumping back and forth.

```suggestion
Serialized in the same way as :ref:`CIS-4-MetadataUrl`.
```

Author: Yes, you're right, it should be a local reference to `MetadataUrl`. I see, however, that there is a repetition of common data types in several standards. Maybe (later) we can organize the standards as a library with common definitions being some sort of `std`

  AuxData ::= (n: Byte²) (data: Byteⁿ)
  RegisterCredentialParameter ::= (credential_info: CredentialInfo) (auxiliary_data: AuxData)

Requirements
~~~~~~~~~~~~

- The credential registration request MUST fail if the credential ID is already present in the registry.
- After successful registration, querying the credential by its ID with :ref:`CIS-4-functions-credentialEntry` MUST succeed.

.. _CIS-4-functions-revokeCredentialIssuer:

``revokeCredentialIssuer``
^^^^^^^^^^^^^^^^^^^^^^^^^^

Revoke a credential by the issuer's request.

.. note::
  This standard does not specify how the issuer is authenticated.
  Implementations can use various mechanisms.
  For example, the transaction sender's address is checked against the issuer's account address stored in the contract.
  Another option is to use ``auxiliary_data`` to implement signature-based authentication mechanism.

Parameter
~~~~~~~~~

The parameter is the credential ID :ref:`CIS-4-CredentialHolderId` and an optional string in the UTF-8 encoding that indicates the revocation reason.

It is serialized as :ref:`CIS-4-CredentialHolderId` followed by 1 byte to indicate whether a reason is included.

> If its value is 0, then no reason string is present, if the value is 1 then the bytes corresponding to the reason string follow::

Reviewer 5: The auxiliary data part of the parameter is left out here. Is that on purpose?

Author: Good catch, `AuxData` should be in the explanation.

    - The credential status is not one of ``Active`` or ``NotActivated`` (see :ref:`CIS-4-functions-credentialStatus`).
    - The signature was intended for a different contract.
    - The signature was intended for a different entrypoint.
    - The signature is expired.
    - The signature can not be validated.
      The smart contract logic SHOULD practice its best efforts to ensure that only the revocation authority can generate and authorize a revocation request with a valid signature.

.. _CIS-4-functions-registerRevocationKeys:

``registerRevocationKeys``
^^^^^^^^^^^^^^^^^^^^^^^^^^

Register public keys that can be used by revocation authorities.

.. note::
  This standard does not specify how the issuer is authenticated.
  Implementations can use various mechanisms.
  For example, the transaction sender's address is checked against the issuer's account address stored in the contract.
  Another option is to use ``auxiliary_data`` to implement signature-based authentication mechanism.

Parameter
~~~~~~~~~

It is serialized as First 2 bytes encode the length (``n``) of the vector of keys, followed by this many :ref:`CIS-4-PublicKeyEd25519` keys::

> 

Reviewer 5: And here

  AuxData ::= (n: Byte²) (data: Byteⁿ)
  RegisterPublicKeyParameters ::= (n: Byte²) (key: PublicKeyEd25519)ⁿ (auxiliary_data: AuxData)

Requirements
~~~~~~~~~~~~

- The revocation MUST fail if some of the keys are already registered.
- The smart contract MUST prevent resetting the nonce associated with a public key.
  For example, the contract logic could keep track of all keys seen by the contract and avoid reusing the same keys even after the keys were made unavailable by calling :ref:`CIS-4-functions-removeRevocationKeys`.

.. _CIS-4-functions-removeRevocationKeys:

``removeRevocationKeys``
^^^^^^^^^^^^^^^^^^^^^^^^

Make a list of public keys unavailable to revocation authorities.

.. note::
  This standard does not specify how the issuer is authenticated.
  Implementations can use various mechanisms.
  For example, the transaction sender's address is checked against the issuer's account address stored in the contract.
  Another option is to use ``auxiliary_data`` to implement signature-based authentication mechanism.

Parameter
~~~~~~~~~

> It is serialized as: First 2 bytes encode the length (``n``) of the vector of keys, followed by this many :ref:`CIS-4-PublicKeyEd25519` keys::

Reviewer 5: Also here the aux data is left out (ignore if it's on purpose)

Parameter
~~~~~~~~~

The parameter is the credential ID :ref:`CIS-4-CredentialHolderId` and an optional string in the UTF-8 encoding that indicates the revocation reason.

It is serialized as :ref:`CIS-4-CredentialHolderId` followed by 1 byte to indicate whether a reason is included.
If its value is 0, then no reason string is present, if the value is 1 then the bytes corresponding to the reason string follow::

  OptionalReason ::= (0: Byte)
                   | (1: Byte) (n: Byte) (reason_string: Byteⁿ)

>   RevokeCredentialIssuerParam ::= (credential_id: CredentialHolderId) (reason: OptionReason)

Reviewer: ```suggestion
  RevokeCredentialIssuerParam ::= (credential_id: CredentialHolderId) (reason: OptionalReason)
```

  RevocationDataOther ::= (credential_id: CredentialHolderId) (signing_data: SigningData) (revocation_key: PublicKeyEd25519) (reason: OptionReason)
  RevokeCredentialHolderParam ::= (signature: SignatureEd25519) (data : RevocationDataOther)

Requirements
~~~~~~~~~~~~

- If revoked successfully, the credential status MUST change to ``Revoked`` (see :ref:`CIS-4-functions-credentialStatus`).
- The message to be signed is produced in the following way:
    - Start with the bytes of the domain separation string ``WEB3ID:REVOKE``.
    - Append ``RevocationDataOther`` bytes from the input parameter.
- The ``RevokeCredentialOtherParam``'s ``signing_data`` MUST include a nonce to protect against replay attacks.
  The revocation authority's nonce is sequentially increased every time a revocation request is successfully executed.
  The function MUST only accept a ``RevokeCredentialOtherParam`` if it has the next nonce following the sequential order.

> - The revocation MUST fail if any of the following conditions are met any of the following conditions are met:

Reviewer: ```suggestion
- The revocation MUST fail if any of the following conditions are met:
```

  RegisterCredentialEvent ::= (249: Byte) (data: CredentialEventData)

``RevokeCredentialEvent``
^^^^^^^^^^^^^^^^^^^^^^^^^

A ``RevokeCredentialEvent`` event MUST be logged when a credential is revoked.
The event records the credential identifier, who requested the revocation (the holder, the issuer or a revocation authority), and an optional string with a short comment on the revocation reason.

The ``RevokeCredentialEvent`` event is serialized as: first a byte with the value of 254, followed by :ref:`CIS-4-CredentialHolderId` (``crednetial_id``), a ``revoker``, and an optional revocation reason (``reason``), serialized similarly to :ref:`CIS-4-functions-revokeCredentialIssuer`; ``revoker`` is serialized as 1 byte to indicate who sent the revocation request ( 0 - issuer, 1 - holder, 2 -revocation authority); if the first byte is 2, then it is followed by a public key :ref:`CIS-4-PublicKeyEd25519` of the revoker::

  Revoker ::= (0: Byte)                         // Issuer
            | (1: Byte)                         // Holder
            | (2: Byte) (key: PublicKeyEd25519) // Other

>   RevokeCredentialEvent ::= (248: Byte) (credential_id: CredentialHolderId) (revoker: Revoker) (reason: OptionReason)

Reviewer: ```suggestion
  RevokeCredentialEvent ::= (248: Byte) (credential_id: CredentialHolderId) (revoker: Revoker) (reason: OptionalReason)
```
